### DISCUSSION TITLE
Version 2: Adding languages to Bible Linker

### DISCUSSION THREAD

**Discussion by majal** [majal/bible-linker-google-docs#24] [Ideas]
Adding languages to Bible Linker is good.

This requires suggestions in the way the code is made and how the Menu will be presented.

Please feel free to **add your suggestions** and **help test** proposed codes in this thread.

**Comment by majal** [comment]
The [first code submission](https://github.com/majal/bible-linker-google-docs/pull/20) and discussion was for the Tagalog language. This was submitted by @JerwellSavella.

**Comment by majal** [comment]
The next was a [request for the Spanish language](https://github.com/majal/bible-linker-google-docs/discussions/21) by @SanSon97.

In response, @venjot [created a refactor](https://github.com/venjot/bible-linker-google-docs/commit/0514a434198e0797e7832bbb253c583187f9a792) that included the Tagalog and Spanish languages.

We will appreciate testing of this new code in the different languages it supports (three for the moment) and any suggestions to improve it.

**Reply by majal** [reply]
Hi @venjot! Thanks for the code. This looks nice. There is just this little bug in the English language.

![image](https://i.imgur.com/tL9nigK.png)

**Reply by majal** [reply]
@venjot, do you think we should add languages for the menu? The phrases "Link verses using..." and "Choose language and version" are in English. 🙂 Maybe later down the line?

**Reply by venjot** [reply]
Hi kuya @majal , I just fixed the duplicate New World Translation (NWL)(WOL).

[Duplicate fix](https://github.com/venjot/bible-linker-google-docs/commit/111ce2c2b5932bf9aa2307b343f3e30bb6b04bd8)

Yes, i think adding languages for the menu is a good feature. :)

**Comment by SanSon97** [comment]
Dear brothers, thank you so much for your time and help.

Although I don't have programming knowledge, I have used Visual Studio Code to translate to Spanish parts of other christian apps. I will be more than happy to translate any parts for you.

I already have your code in VSC.

SanSon97

**Reply by majal** [reply]
Awesome. We'll surely ask for your help in Spanish translation. 🙂

While the language additions is in development, do you think you could help us test out the Spanish version from @venjot’s development code?

https://github.com/venjot/bible-linker-google-docs/tree/feature/add-bible-languages

**Comment by SanSon97** [comment]
Thank you so much @majal and @venjot for the opportunity to help. I have done a couple testings, it works great!!
Tried a long article and only got a few errors as seen in the video.

https://user-images.githubusercontent.com/102166309/179379469-dd69d378-fc07-4939-9fd9-8a1d53ab00ca.mp4 

One question. Is there a way to open the links in JW Library? The link in my iPad and Windows PC opens JW Library but then goes straight to wol.jw.org

**Reply by venjot** [reply]
Hi @SanSon97 ,
Can you please send me the long article that you used for testing?  So we can also know what are the other abbreviation that we need to add for bible books.

as of now, We used the abbreviation we found in the Bible. We will add "Isa" as part of abbreviation for Isaias. as of now, we only used "Is" for abbreviation for "Isaias". We will update it.

![Screenshot from 2022-07-17 09-01-09](https://user-images.githubusercontent.com/3022302/179379709-0c581cf9-0f8f-43b0-b160-b9bd6b6bb5fd.png)

Regarding your question, 

> Is there a way to open the links in JW Library? The link in my iPad and Windows PC opens JW Library but then goes straight to wol.jw.org.

We will check if it it is possible.

Thanks for your help :)

**Reply by SanSon97** [reply]
Test 2.
The link goes to the English Bible, notice that I had to change the language link from 'ES' to only 'S' and it fixed the issue

https://user-images.githubusercontent.com/102166309/179380328-488da342-9210-4a32-a9ca-30fcee37fd04.mp4

e

**Reply by SanSon97** [reply]
@venjot Absolutely, here is the Doc without links 
[Jehova.docx](https://github.com/majal/bible-linker-google-docs/files/9126896/Jehova.docx)
.

**Reply by venjot** [reply]
Hi @SanSon97 , You might not using the latest code? the latest link for Spanish, We've changed it from "ES" to "S".

see our latest code change.

![Screenshot from 2022-07-17 09-59-24](https://user-images.githubusercontent.com/3022302/179380811-839b2624-4d32-47f8-a31e-dffd7c66a434.png)

**Reply by venjot** [reply]
Hi @SanSon97 ,

Can you help me with this? We've noticed the bible verse below.

Berajot 9:5.

May I know what bible book is that in english?

We've found the following as missing abbreviation in spanish.
- Rev
- Isa

Thanks you :) 
We really appreciate your help.

**Reply by SanSon97** [reply]
That's the name of of a Mishnah book :) The article is making reference to it.

Here's the English paragraph
*** it-2 p. 6 Jehovah ***
Yet, despite these negative views, one also finds in the first section of the Mishnah the positive injunction that “a man should salute his fellow with [the use of] the Name [of God],” the example of Boaz (Ru 2:4) then being cited.—Berakhot 9:5.

**Reply by SanSon97** [reply]
BTW it's my pleasure being able to help.

**Reply by venjot** [reply]
> That's the name of of a Mishnah book :) The article is making reference to it.
> 
> Here's the English paragraph *** it-2 p. 6 Jehovah *** Yet, despite these negative views, one also finds in the first section of the Mishnah the positive injunction that “a man should salute his fellow with [the use of] the Name [of God],” the example of Boaz (Ru 2:4) then being cited.—Berakhot 9:5.

I see.

We've already added "Isa" as abbreviation for "Isaias" and "Rev" as abbreviation for "Apocalipsis".

see latest code: [Latest Code](https://github.com/venjot/bible-linker-google-docs/blob/feature/add-bible-languages/Code.gs)

**Comment by JerwellSavella** [comment]
Thanks for the code Kuya @venjot! , I submitted a PR to the feature/add-bible-languages branch to add [Tagalog Abbreviation](https://github.com/venjot/bible-linker-google-docs/pull/1/commits/8b89d837d6af5372793dab090f09e954f9771529) as well. I will also do some tests on the latest code. Thanks po!

**Reply by venjot** [reply]
Hi @JerwellSavella , I have approved the pull request :) thanks for adding abbreviation for tagalog.

**Comment by SanSon97** [comment]
Thank you all for making the app available in Spanish.

Regarding JW Library, the issue is not in the link, I found the way to opening the link in the app.

![image](https://user-images.githubusercontent.com/102166309/179402801-b5ff6fed-784b-4970-be08-2ed203826ec2.png)

**Reply by venjot** [reply]
@SanSon97 That's great to hear. Thanks for the info. :)

**Reply by venjot** [reply]
Hi @SanSon97 , 

how's your testing for Spanish? if ever you encounter a bible book name or abbreviation that was not recognized by the bible linker in Spanish, please let us know so we can add/update it. 

Thank you :)

**Reply by SanSon97** [reply]
Gretings @venjot ,
Thanks for asking. I have found older abbreviations. Here are some from this article  *** w85 15/3 pág. 15 En unidad con el Creador de la organización universal ***
![image](https://user-images.githubusercontent.com/102166309/180197866-08e5b8db-24a0-415d-a350-a159b3e7fde7.png)

In those errors it did create the links but got confused because the text didn't end with an ) I guess?

As you can see in the document, tried to detect 'is' and 'os' when parsing Apocalipsis and Efesios.

Revelación and Salmo need to be added.

Thanks a lot. :) 
[w88 15_3 pg. 15 Test.docx](https://github.com/majal/bible-linker-google-docs/files/9158790/w88.15_3.pg.15.Test.docx)

**Reply by SanSon97** [reply]
Will keep on trying more articles.

**Reply by SanSon97** [reply]
Found this ones on this week's study article:

Salmo= Sal.
Proverbios= Prov.
Ezequiel= Ezeq.
Mateo = Mat.
Lucas = Luc.
Hechos= Hech.
Romanos= Rom.
Efesios = Efes.
2 Tesalonicenses= 2 Tes.
Hebreos= Heb.
Apocalipsis = Apoc.

**Reply by venjot** [reply]
Hi @SanSon97 ,
Thank you for this. I will update the abbreviation.

Can you also update the script that you are using? using this [Latest Code](https://github.com/venjot/bible-linker-google-docs/blob/feature/add-bible-languages/Code.gs)

I'm not sure if I used the correct bible for my reference previously, I used the bible text below. it seems in my reference, it is "Salmos" not "Salmo". also, Apocalipsis not "Revelación". 

![image](https://user-images.githubusercontent.com/3022302/180343721-aa4cbb0a-6b8b-4263-b1b6-f306c55a3e91.png)

but it seems that Apocalipsis and Revelación are both used in Spanish. is Salmos and Salmo both used in Spanish as well?

**Reply by SanSon97** [reply]
Dear @venjot I did copy the code but got this error in line 353
![image](https://user-images.githubusercontent.com/102166309/180764691-6937f03c-f882-441f-87b9-f4f04e0ad9fe.png)

**Reply by venjot** [reply]
> Dear @venjot I did copy the code but got this error in line 353 ![image](https://user-images.githubusercontent.com/102166309/180764691-6937f03c-f882-441f-87b9-f4f04e0ad9fe.png)

Hi @SanSon97 , May I know what bible you are using and the document you are testing? as I was not able to replicate the issue. 

Thank you so much :)

**Reply by venjot** [reply]
Hi @SanSon97 , I was able to replicate the issue now.

I think you run the AppScript while the "getBibleLink" is selected.
![image](https://user-images.githubusercontent.com/3022302/180900434-cdc1307e-dbec-435c-90ee-83621bad979c.png)


Can you please try to run it while "bibleLinker" is selected? You should not have any issues after running it.

![image](https://user-images.githubusercontent.com/3022302/180900506-a4f10b38-63b7-4cd5-9a99-82c91641aab4.png)

**Comment by SanSon97** [comment]
Hello @venjot you are using the correct and latest abbreviation, it's just that in some publications it varies. Also, some of those are from the Rbi8. Here is a picture of the 1987 Rbi8.
I'll keep checking for new/different ones.
![image](https://user-images.githubusercontent.com/102166309/180760243-8a15937e-ccdf-454f-b273-0c0469d83735.png)

**Reply by venjot** [reply]
Hi @SanSon97 , thanks for this. :) I will let you know once I have updated the code for the book names and abbreviations. :)

**Reply by SanSon97** [reply]
Thank you so much @venjot I noticed only a few differences in both Indexes, the other are new abbreviations found in the newer Magazines. Like this week's study article:
![image](https://user-images.githubusercontent.com/102166309/181031871-10cc4863-4a46-4dfa-b3fd-ddfcf484f4a9.png)

**Reply by venjot** [reply]
Hi @SanSon97 , We've updated the abbreviation for Spanish.

see the latest code. [Code.gs](https://raw.githubusercontent.com/venjot/bible-linker-google-docs/feature/add-bible-languages/Code.gs)

We've found an issue that we haven't fixed. the issue is it is using "–" rather than "-"
Example:
Génesis 6:1–7:23

We will update you once we fixed it. :)

**Reply by venjot** [reply]
Hi @SanSon97 , We've already applied the fix for the issue that bible verse is using "–" rather than "-".

Example:
Génesis 6:1–7:23

see latest code: [Code.gs](https://raw.githubusercontent.com/venjot/bible-linker-google-docs/feature/add-bible-languages/Code.gs)

**Comment by majal** [comment]
Hi everyone!

I had some time to fiddle with the code again after a year. Version 2 is coming.

The main difference with v1 is that all data on v2 is now externally linked via JSON. This will make it easy to customize the Linker for other languages.

All core functions have been rewritten from scratch. Of course it needs testing to see if there are bugs or edge cases.

TO DO:
 - Make selections work
 - Parse single-chapter Bible verses
 - Rewrite the UI

Code:
https://github.com/majal/bible-linker-google-docs/commits/main/Code.gs

Initial data for English:
https://github.com/majal/bible-linker-google-docs/commits/main/bible-data/en_jw.json
